Invalid Largest Reference is a stream error.

bencebeky opened the pull request:
If Largest Reference does not figure, then the decoded Base Index will probably be bogus, resulting in corrupt entries.  I think it's better to close the connection with an error.

bencebeky commented on draft-ietf-quic-qpack.md:
```diff
@@ -825,22 +825,32 @@ The decoder reconstructs the Largest Reference using the following algorithm:
-      LargestReference += TotalNumberOfInserts - CurrentWrapped
+      LargestReference += TotalNumberOfInserts
+
+      if LargestReference <= CurrentWrapped:
```
I put the error checking here because I am using unsigned integers.  If the majority of implementors are expected to use signed integers (after all, 2**62 required by the spec can comfortably be represented using 64 bit signed integers), then the original `LargestReference += TotalNumberOfInserts - CurrentWrapped` line can be kept and the error checking woul look like:

```
   if LargestReference <= 0:
      Error
```

martinthomson commented:
This will conflict with #2248.

bencebeky commented:
> This will conflict with #2248.

No worries, I do not expect all of my PRs to be approved (even though I think they are all improvements).  I'm just drafting them to start a discussion, because for small things like this it is much easier to talk about an actual PR then to try describing the changes.

If there is consensus to merge either this or #2248, I'm happy to rebase the other one.

MikeBishop commented:
I'm confused -- if LR is invalid, the HEADERS frame can't be interpreted, I agree.  It's unclear why a failure to decode anything on a request stream could constitute an encoder stream error.

bencebeky commented:
> I'm confused -- if LR is invalid, the HEADERS frame can't be interpreted, I agree. It's unclear why a failure to decode anything on a request stream could constitute an encoder stream error.

Good point.  Let's make it a stream error (with the usual code HTTP_QPACK_DECOMPRESSION_FAILED) then.

bencebeky commented:
I rebased on #2111, incorporated the suggestion by @MikeBishop above to make it a stream error instead of a connection error, and reconciled that sometimes InsertCount and sometimes ReqInsertCount was used in the decoding algorithm.

afrind commented:
overtaken by events
